waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,512 @@
+
+      query.paginationCursor = response.paginationCursor
+
+      for kv in response.messages:
+        let handleRes = catch:
+          await wakuArchive.handleMessage(kv.pubsubTopic.get(), kv.message.get())
+
+        if handleRes.isErr():
+          error "message transfer failed", error = handleRes.error.msg
+          # Messages can be synced next time since they are not added to storage yet.
+          continue
+
+        self.messageIngress(kv.pubsubTopic.get(), kv.messageHash, kv.message.get())
+
+      if query.paginationCursor.isNone():
+        break
+
+    return ok()
+
+  return ok(callback)
+
+proc initFillStorage(
+    self: WakuSync, wakuArchive: WakuArchive
+): Future[Result[void, string]] {.async.} =
+  if wakuArchive.isNil():
+    return err("waku archive unavailable")
+
+  let endTime = getNowInNanosecondTime()
+  let starTime = endTime - self.syncRange.nanos
+
+  var query = ArchiveQuery(
+    includeData: true,
+    cursor: none(ArchiveCursor),
+    startTime: some(starTime),
+    endTime: some(endTime),
+    pageSize: 100,
+  )
+
+  while true:
+    let response = (await wakuArchive.findMessages(query)).valueOr:
+      return err($error)
+
+    for i in 0 ..< response.hashes.len:
+      let hash = response.hashes[i]
+      let topic = response.topics[i]
+      let msg = response.messages[i]
+
+      self.messageIngress(topic, hash, msg)
+
+    if response.cursor.isNone():
+      break
+
+    query.cursor = response.cursor
+
+  return ok()
+
+proc new*(
+    T: type WakuSync,
+    peerManager: PeerManager,
+    maxFrameSize: int = DefaultMaxFrameSize,
+    syncRange: timer.Duration = DefaultSyncRange,
+    syncInterval: timer.Duration = DefaultSyncInterval,
+    relayJitter: Duration = DefaultGossipSubJitter,
+    wakuArchive: WakuArchive,
+    wakuStoreClient: WakuStoreClient,
+    pruneCallback: Option[PruneCallback] = none(PruneCallback),
+    transferCallback: Option[TransferCallback] = none(TransferCallback),
+): Future[Result[T, string]] {.async.} =
+  let storage = NegentropyStorage.new().valueOr:
+    return err("negentropy storage creation failed")
+
+  var sync = WakuSync(
+    storage: storage,
+    peerManager: peerManager,
+    maxFrameSize: maxFrameSize,
+    syncInterval: syncInterval,
+    syncRange: syncRange,
+    relayJitter: relayJitter,
+    pruneOffset: syncInterval div 100,
+  )
+
+  sync.initProtocolHandler()
+
+  sync.pruneCallBack = pruneCallback
+
+  if sync.pruneCallBack.isNone():
+    let res = sync.createPruneCallback(wakuArchive)
+
+    if res.isErr():
+      error "pruning callback creation error", error = res.error
+    else:
+      sync.pruneCallBack = some(res.get())
+
+  sync.transferCallBack = transferCallback
+
+  if sync.transferCallBack.isNone():
+    let res = sync.createTransferCallback(wakuArchive, wakuStoreClient)
+
+    if res.isErr():
+      error "transfer callback creation error", error = res.error
+    else:
+      sync.transferCallBack = some(res.get())
```

Comment:
The following seems not needed:
```suggestion
  if wakuArchive.isNil():
    return err("waku archive unavailable")

  if wakuStoreClient.isNil():
    return err("waku store client unavailable")
```

Reply:
We can't error in case archive and store client are nil otherwise we would have to include those in ours tests.